### PR TITLE
Fixed Dissolution Date bug

### PR DESCRIPTION
**Pull request by JazzarKarim**
*Issue #:* /bcgov/entity#18008

*Description of changes:*
- Fixed the bug in which the UI doesn't let the user select a dissolution date that is the same as the registration date.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

File: src/rules/date-rules.ts [hunk 1]
```diff
@@ -15,6 +15,7 @@ export const DateRuleHelpers = {
   isBetweenDates (minDate: Date, maxDate: Date, dateStrToValidate: string): boolean {
     if (!dateStrToValidate) return true
     const date = new Date(dateStrToValidate) // eg, September 5, 2022
+    minDate.setHours(0, 0, 0) // Removing the hour to properly compare in case selected date was same as minimum.
     return (date >= minDate) && (date <= maxDate)
```
Comment (JazzarKarim) [inline src/rules/date-rules.ts:19]: There was a bit of weird behavior here. When selecting the date as same as registration date, the error (rule) was being shown:
rule is shown

Apparently, the issue is with the hour:
weird dates
The first line is the `date` while the second one is the `minDate`. Hence, the rule is being triggered.

Setting the hour to be the same (`00:00:00`) fixes this issue.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-create-dev--pr-577-3s0yhst0.web.app

**Comment by JazzarKarim** [comment]
For testing:
- https://business-create-dev--pr-577-3s0yhst0.web.app/dissolution-firm?id=FM1052377&accountid=3040
- https://business-create-dev--pr-577-3s0yhst0.web.app/dissolution-firm?id=FM1056109&accountid=3040

File: src/rules/date-rules.ts (hunk 1 above)
Comment (severinbeauvais) [reply src/rules/date-rules.ts:19]: I think this is indeed the root cause 👍 

What do you think about fixing maxDate similarly (and also isNotBeforeDate() and isNotAfterDate()) ?

File: src/views/DissolutionFirm/DissolutionFirm.vue [hunk 2]
```diff
@@ -58,7 +58,7 @@
         <h2>Business Dissolution Date</h2>
         <p class="mt-4 ">
           Enter the dissolution date of the business.
-          The dissolution date must be after the business start date and registration date.
+          The dissolution date must be on or after the registration date.
```
Comment (severinbeauvais) [inline src/views/DissolutionFirm/DissolutionFirm.vue:61]: Did Mihai approve this business logic change?

Does the Legal API need to validate this as well?

Could you please add some unit tests to verify the 3 date rule helpers with these edge cases?

Comment (JazzarKarim) [reply src/views/DissolutionFirm/DissolutionFirm.vue:61]: Riyaz opened the issue while Mihai had a look at it and did some changes in the ticket. So, I assumed that he already had a look at it and is OK with it. I'll double check and speak with him though Sev.

I don't think so as I don't know why, but the FE had an extra `+ 1` for the day here. So, I think it should be fine.

Sure thing! On it.

Comment (severinbeauvais) [reply src/views/DissolutionFirm/DissolutionFirm.vue:61]: Were you able to file a firm dissolution on the same date as the registration date? That would prove out the Legal API :)

Comment (severinbeauvais) [reply src/views/DissolutionFirm/DissolutionFirm.vue:61]: The `+1` was probably to make the day after (ie, after the business start date).

Comment (JazzarKarim) [reply src/views/DissolutionFirm/DissolutionFirm.vue:61]: Yup! I tested it.
confirm date

File: src/rules/date-rules.ts [hunk 3]
```diff
@@ -16,18 +16,21 @@ export const DateRuleHelpers = {
     if (!dateStrToValidate) return true
     const date = new Date(dateStrToValidate) // eg, September 5, 2022
     minDate.setHours(0, 0, 0) // Removing the hour to properly compare in case selected date was same as minimum.
+    maxDate.setHours(0, 0, 0)
```
Comment (severinbeauvais) [inline src/rules/date-rules.ts:19]: Shouldn't maxDate be set to 23:59:59.999 ?

File: tests/unit/DissolutionFirm.spec.ts [hunk 4]
```diff
@@ -138,5 +138,22 @@ for (const test of dissolutionFirmTestCases) {
...
+      store.setCurrentJsDate(new Date('2022-06-14T12:00:00.000Z'))
+
+      expect(rules[0]('')).toBe('Dissolution date is required') // no date is selected
+      expect(rules[0]('October 16, 2023')).toBe(true) // date is selected
+      // A date before the registration date is selected (invalid)
+      expect(rules[1]('June 5, 2022')).toContain('Dissolution Date must be after June 6, 2022 and up to')
+      expect(rules[1]('June 5, 2022')).toContain('June 14, 2022')
+      // A valid date is selected (after registration date and not in the future)
+      expect(rules[1]('June 7, 2022')).toBe(true)
+    })
```
Comment (severinbeauvais) [inline tests/unit/DissolutionFirm.spec.ts:157]: Can you test the edge cases, too, please?

Example:  "Jun 5 0:00:00 >= Jun 5" and "Jun 6 23:59:59 <= Jun 6"?

Comment (JazzarKarim) [reply tests/unit/DissolutionFirm.spec.ts:157]: On it!

Comment (severinbeauvais) [reply tests/unit/DissolutionFirm.spec.ts:158]: What about the other edge? (future date)

File: src/rules/date-rules.ts (hunk 3 above)
Comment (severinbeauvais) [reply src/rules/date-rules.ts]: There's another one on line 33.

File: tests/unit/DissolutionFirm.spec.ts (hunk 4 above)
Comment (JazzarKarim) [reply tests/unit/DissolutionFirm.spec.ts:161]: Done haha. Was exactly thinking that. Sorry about that Sev.

Comment (JazzarKarim) [reply tests/unit/DissolutionFirm.spec.ts:161]: Sev, the tests are running locally but failing here. I can't seem to figure out why. Could it be related to timezones?

Comment (JazzarKarim) [reply tests/unit/DissolutionFirm.spec.ts:163]: Never mind, I fixed it 😁 My suspicions were correct, it was timezone issue. Should be fine now!

Comment (seeker25) [reply tests/unit/DissolutionFirm.spec.ts:163]: Timezone issues 🤯

**Review by severinbeauvais — APPROVED** [review]
Looks good.

**Review by PaulGarewal — APPROVED** [review]
LGTM 👍 Great job resolving all the comments 😃

**Comment by JazzarKarim** [comment]
Merging now after having a long discussion with Mihai regarding this. Should be good!
